### ISSUE TITLE
feat: add fold4d (col2im) operation as inverse of unfold4d

### ISSUE THREAD

**Issue by antimora** [tracel-ai/burn#4519]
## Summary

Burn has `unfold4d` (im2col) but no corresponding `fold4d` (col2im) operation. This is needed for implementing the ONNX `Col2Im` operator efficiently in burn-onnx, and is generally useful for any operation that needs to rearrange column blocks back into a multidimensional image (e.g., custom convolution implementations, attention mechanisms with spatial unfolding).

## Motivation

- PyTorch has both [`unfold`](https://pytorch.org/docs/stable/generated/torch.nn.Unfold.html) and [`fold`](https://pytorch.org/docs/stable/generated/torch.nn.Fold.html) as paired operations
- ONNX has both `Im2Col` (mapped to `unfold4d`) and `Col2Im` (no Burn equivalent)
- Without a native `fold4d`, burn-onnx has to generate per-element loop code with `slice`/`slice_assign`, which is orders of magnitude slower than a batched tensor operation (see tracel-ai/burn-onnx#169)
- Burn already has an internal `col2im` kernel in `burn-cubecl` for `conv_transpose2d`, but it's not exposed as a public tensor operation

## Proposed API

```rust
/// Rearranges columns back into a multidimensional image (inverse of unfold4d).
///
/// # Shapes
/// * x:      [batch_size, channels * kernel_h * kernel_w, num_blocks]
/// * returns: [batch_size, channels, height, width]
pub fn fold4d<B: Backend>(
    x: Tensor<B, 3>,
    output_size: [usize; 2],
    kernel_size: [usize; 2],
    options: UnfoldOptions,  // reuse existing options (stride, padding, dilation)
) -> Tensor<B, 4>
```

Mirrors PyTorch's `torch.nn.Fold` signature. Reuses the existing `UnfoldOptions` struct since the parameters are identical.

## Workaround / Default Implementation

A default implementation can be built using `scatter` with `IndexingUpdateOp::Add`:

1. Compute the index mapping from column positions to output image positions (accounting for stride, dilation, padding)
2. Use `Tensor::scatter(dim, indices, values, IndexingUpdateOp::Add)` to accumulate overlapping blocks into the output tensor

This avoids per-element loops entirely and works on any backend. Backend-specific implementations (e.g., reusing the existing cubecl `col2im` kernel) can override for better performance.

## Context

- burn-onnx PR implementing Col2Im: tracel-ai/burn-onnx#169
- Existing `unfold4d`: `crates/burn-tensor/src/tensor/module.rs`
- Existing internal `col2im` kernel: `crates/burn-cubecl/src/kernel/conv/conv_transpose2d/col2im.rs`

**Comment by Capataina** [comment]
Hey, this looks like something I might be able to help with. I've got some experience with ONNX graphs from prior work. I'd like to pick this up.

**Comment by Capataina** [comment]
I went through the issue and the surrounding code, ready to start. Two scoping things I'd want to lock down before opening the PR:

1. Default-impl strategy, scatter-add or conv_transpose2d with an identity weight? The issue recommends scatter, that's the direct path. conv_transpose2d would mirror how unfold4d_using_conv2d works structurally, and burn-cubecl's conv_transpose2d already routes through a col2im kernel for the GEMM path. Either works mathematically, leaning scatter unless you prefer the conv_transpose2d approach.

2. Signature, output_size as a separate parameter (matching torch.fold) or inside an options struct? Leaning separate since output_size is required (no sensible default) and that matches the issue body's example.

I can also do the unfold4d autodiff TODO at crates/burn-autodiff/src/ops/module.rs:1314-1327 while I'm at it.

Let me know if either of those calls is obvious, otherwise I'll go with scatter and separate output_size.

**Comment by laggui** [comment]
1. The default implementation can use scatter add
2. It can go in the `FoldOptions` struct even if there is no default. It's part of the fold operation options/args.

**Comment by Capataina** [comment]
Just to let you know, I'm still working on this; just taking some time

**Comment by charitarthchugh** [comment]
Hi @Capataina — friendly heads-up on a possible overlap. I picked up #786 , which asks for torch.nn.Fold support — which looks like the same operation you're implementing here as fold4d/col2im. If you are still working on this, I am happy to stay out of your way (or help/review if useful). If it's stalled and you'd rather hand it off, I'm glad to take it forward. Thanks!

**Comment by Capataina** [comment]
@charitarthchugh Hey, you can take the issue; no worries. Life happened so I had to step back from this one sadly!

**Comment by antimora** [comment]
Done in #5173 (`fold4d` / `Fold4d`). The performance follow-up is #5782. Closing.
